src/map/mob.c

```diff
@@ -3860,6 +3901,209 @@ static inline int mob_parse_dbrow_cap_value(int class_, int min, int max, int va
 	return value;
 }
 
+/**
+ * Reads one possible option for a option slot in a option drop group
+ * @param option : Libconfig entry
+ * @param entry : memory db entry for current slot
+ * @param idx : index of entry where this option should be inserted at
+ * @param calc_rate : if rates should be recalculated after reading all entries
+ * @param slot : option group slot being read (for messages)
+ * @param group : option group being read (for messages)
+ * @return true if it successfully read the entry, false otherwise
+ */
+static bool mob_read_optdrops_option(struct config_setting_t *option, struct optdrop_group_optslot *entry, int *idx, bool *calc_rate, int slot, const char *group)
+{
+	nullpo_retr(false, option);
+	nullpo_retr(false, entry);
+	nullpo_retr(false, idx);
+	nullpo_retr(false, calc_rate);
+	nullpo_retr(false, group);
+
+	const char *name = config_setting_name(option);
+	int opt_id;
+
+	if (strncmp(name, "Rate", 4) == 0)
+		return true;
+
+	if (script->get_constant(name, &opt_id) == false) {
+		ShowWarning("mob_read_optdrops_option: Invalid option \"%s\" for option slot %d of %s group, skipping.\n", name, slot, group);
+		return false;
+	}
+	
+	int min = 0, max = 0, opt_rate = 0;
+	if (config_setting_is_number(option)) {
+		// OptionName: value
+		min = libconfig->setting_get_int(option);
+	} else if (config_setting_is_array(option)) {
+		// OptionName: [min, max]
+		// OptionName: [min, max, rate]
+		int slen = libconfig->setting_length(option);
+		
+		if (slen >= 2) {
+			// [min, max,...]
+			min = libconfig->setting_get_int_elem(option, 0);
+			max = libconfig->setting_get_int_elem(option, 1);
+		}
+		
+		if (slen == 3) {
+			// [min, max, rate]
+			opt_rate = libconfig->setting_get_int_elem(option, 2);
+		}
+	} else {
+		ShowWarning("mob_read_optdrops_option: Invalid value \"%s\" for option slot %d of %s group, skipping.\n", name, slot, group);
+		return false;
+	}
+
+	if (max < min)
+		max = min;
+	
+	entry->options[*idx].id = opt_id;
+	entry->options[*idx].min = min;
+	entry->options[*idx].max = max;
+	entry->options[*idx].rate = opt_rate;
+
+	if (entry->options[*idx].rate == 0)
+		*calc_rate = true;
+
+	(*idx)++;
+
+	return true;
+}
+
+/**
+ * Reads the settings for one random option slot of a random option drop group.
+ * @param optslot : The slot entry from config file
+ * @param n : slot index
+ * @param group_id : Group index
+ * @param group : group name (used in messages)
+ * @return true if it succesfully read, false otherwise
+ */
+static bool mob_read_optdrops_optslot(struct config_setting_t *optslot, int n, int group_id, const char *group)
+{
+	nullpo_retr(false, optslot);
+	nullpo_retr(false, group);
+
```

Comment:
need add here Assert_ret for group_id or check group_id before used as array index

Reply:
same for n